### PR TITLE
Added earlier validation for command line arguments

### LINKED ISSUE #797: Check if command line arguments are valid
At the moment the `esmvaltool` command does not check if any invalid arguments are provided, e.g. a command like this
```
esmvaltool run --config-flie config-user.yml examples/recipe_python.yml
```
will happily run and try to use the configuration file in default location `~/.esmvaltool/config-user.yml` instead of the one provided on the command line. This is very confusing if that file contains an error and you do not spot the typo in the command.

### PR DESCRIPTION
**Pull request by schlunma**
## Description

This PR adds additional validation for the validity of command line arguments before reading the configuration. This allows us to detect errors like

```
> esmvaltool run --config-usa ~/config-user.yml ~/tmp/recipes/recipe_cox18nature.yml
...
esmvalcore.experimental.config._validated_config.InvalidConfigParameter: `config-usa` is not a valid config parameter.
```

In addition, this also catches wrong types (this has also been caught previously, but later in the run):

```
> esmvaltool run --config-file ~/config/DKRZ_default.yml --offline=aaaa ~/ESMValTool/esmvaltool/recipes/examples/recipe_python.yml
...
esmvalcore.experimental.config._validated_config.InvalidConfigParameter: Key `offline`: Could not convert `aaaa` to `bool`
```

Moreover, I renamed the option `skip-nonexistent` to `skip_nonexistent` across the entire tool. This is now consistent with other options. `Fire` automatically renames `-` to `_` anyway, so both spelling options are (still) fine in the command line.

Closes #797

***

## [Before you get started](https://docs.esmvaltool.org/projects/ESMValCore/en/latest/contributing.html#getting-started)

- [x] [☝ Create an issue](https://github.com/ESMValGroup/ESMValCore/issues) to discuss what you are going to do

## [Checklist](https://docs.esmvaltool.org/projects/ESMValCore/en/latest/contributing.html#checklist-for-pull-requests)

It is the responsibility of the author to make sure the pull request is ready to review. The icons indicate whether the item will be subject to the [🛠 Technical][1] or [🧪 Scientific][2] review.

[1]: https://docs.esmvaltool.org/en/latest/community/review.html#technical-review
[2]: https://docs.esmvaltool.org/en/latest/community/review.html#scientific-review

- [x] [🧪][2] The new functionality is [relevant and scientifically sound](https://docs.esmvaltool.org/projects/ESMValCore/en/latest/contributing.html#scientific-relevance)
- [x] [🛠][1] This pull request has a [descriptive title and labels](https://docs.esmvaltool.org/projects/ESMValCore/en/latest/contributing.html#pull-request-title-and-label)
- [x] [🛠][1] Code is written according to the [code quality guidelines](https://docs.esmvaltool.org/projects/ESMValCore/en/latest/contributing.html#code-quality)
- [x] [🧪][2] and [🛠][1] [Documentation](https://docs.esmvaltool.org/projects/ESMValCore/en/latest/contributing.html#documentation) is available
- [x] [🛠][1] [Unit tests](https://docs.esmvaltool.org/projects/ESMValCore/en/latest/contributing.html#tests) have been added
- [x] [🛠][1] Changes are [backward compatible](https://docs.esmvaltool.org/projects/ESMValCore/en/latest/contributing.html#backward-compatibility)
- [x] [🛠][1] Any changed [dependencies have been added or removed](https://docs.esmvaltool.org/projects/ESMValCore/en/latest/contributing.html#dependencies) correctly
- [x] [🛠][1] The [list of authors](https://docs.esmvaltool.org/projects/ESMValCore/en/latest/contributing.html#list-of-authors) is up to date
- [x] [🛠][1] All [checks below this pull request](https://docs.esmvaltool.org/projects/ESMValCore/en/latest/contributing.html#pull-request-checks) were successful

***

To help with the number pull requests:

- 🙏 We kindly ask you to [review](https://docs.esmvaltool.org/en/latest/community/review.html#review-of-pull-requests) two other [open pull requests](https://github.com/ESMValGroup/ESMValCore/pulls) in this repository

### CODE REVIEW

File: esmvalcore/_main.py [hunk 1]
```diff
@@ -368,12 +368,35 @@ def run(self,
...
         from ._recipe import TASKSEP
         from .cmor.check import CheckLevels
         from .esgf._logon import logon
 
+        # Check validity of optional command line arguments with experimental
+        # API
+        with warnings.catch_warnings():
+            # ignore experimental API warning
+            warnings.simplefilter("ignore")
+            from .experimental.config._config_object import Config
```
Comment (stefsmeets) [inline esmvalcore/_main.py:383]: Avoid name collision with :133

```suggestion
            from .experimental.config._config_object import ExpConfig
```

**Review by stefsmeets — APPROVED** [review]
Nice use of the experimental config validators 👍

File: esmvalcore/_main.py [hunk 2]
```diff
@@ -368,12 +368,35 @@ def run(self,
...
+            'skip_nonexistent': skip_nonexistent,
+            'offline': offline,
+            'diagnostics': diagnostics,
+            'check_level': check_level,
+        }
+        all_optional_kwargs = dict(kwargs)
+        for (key, val) in explicit_optional_kwargs.items():
+            if val is not None:
+                all_optional_kwargs[key] = val
+        Config(all_optional_kwargs)
```
Comment (stefsmeets) [inline esmvalcore/_main.py:398]:
```suggestion
        ExpConfig(all_optional_kwargs)
```

**Comment by schlunma** [comment]
Any further comments on this @zklaus?

**Review by zklaus — APPROVED** [review]
Looking good. Nice improvement, thanks for doing it! :beers:

**Pull request merged**